pkg/base/replicated.go

```diff
@@ -152,7 +157,149 @@ func renderReplicated(u *upstreamtypes.Upstream, renderOptions *RenderOptions) (
+	fileContentsMap := make(map[string][]byte)
+	kotsKinds := kotsutil.EmptyKotsKinds()
+
+	for _, upstreamFile := range upstreamFiles {
+		document := &Document{}
+		if err := yaml.Unmarshal(upstreamFile.Content, document); err != nil {
+			continue
+		}
+
+		decode := scheme.Codecs.UniversalDeserializer().Decode
+		decoded, gvk, err := decode(upstreamFile.Content, nil, nil)
+		if err != nil {
+			if document.APIVersion == "kots.io/v1beta1" || document.APIVersion == "troubleshoot.sh/v1beta2" || document.APIVersion == "velero.io/v1" {
+				return nil, nil, errors.Wrapf(err, "failed to decode kots yaml %s", upstreamFile.Path)
+			}
+			continue
+		}
+
+		switch gvk.String() {
+		case "kots.io/v1beta1, Kind=Config":
+			// Use the rendered Config instead of the upstream.
+			kotsKinds.Config = renderedConfig
+			configBytes, err := yaml.Marshal(renderedConfig)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to marshal rendered config")
+			}
+			fileContentsMap[upstreamFile.Path] = configBytes
+
+		case "kots.io/v1beta1, Kind=ConfigValues":
+			fileContentsMap["configvalues.yaml"] = upstreamFile.Content
+			kotsKinds.ConfigValues = decoded.(*kotsv1beta1.ConfigValues)
+
+		case "kots.io/v1beta1, Kind=HelmChart":
+			helmChart := decoded.(*kotsv1beta1.HelmChart)
+
+			mergedValues := helmChart.Spec.Values
+			if mergedValues == nil {
+				mergedValues = map[string]kotsv1beta1.MappedChartValue{}
+			}
+			for _, optionalValues := range helmChart.Spec.OptionalValues {
+				parsedBool, err := strconv.ParseBool(optionalValues.When)
+				if err != nil {
+					return nil, nil, errors.Wrap(err, "failed to parse when conditional on optional values")
+				}
+				if !parsedBool {
+					continue
+				}
+				if optionalValues.RecursiveMerge {
+					mergedValues = kotsv1beta1.MergeHelmChartValues(mergedValues, optionalValues.Values)
+				} else {
+					for k, v := range optionalValues.Values {
+						mergedValues[k] = v
+					}
+				}
+			}
+
+			helmChart.Spec.Values = mergedValues
+			helmChart.Spec.OptionalValues = nil
+
+			mergedContent, err := yaml.Marshal(helmChart)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed to marshal merged helm chart")
+			}
+
+			renderedChart, err := builder.RenderTemplate(upstreamFile.Path, string(mergedContent))
+
+			decodedChart, _, err := decode([]byte(renderedChart), nil, nil)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "failed to decode rendered helm chart %s", upstreamFile.Path)
+			}
+
+			if kotsKinds.HelmCharts == nil {
+				kotsKinds.HelmCharts = &kotsv1beta1.HelmChartList{}
+			}
+
+			kotsKinds.HelmCharts.Items = append(kotsKinds.HelmCharts.Items, decodedChart.(*kotsv1beta1.HelmChart))
+			fileContentsMap[upstreamFile.Path] = []byte(renderedChart)
+
+		default:
```

Comment:
would this basically trigger for every single file in the upstream? that seems like wasted performance to render and this discard.

Reply:
yeah, fixed now. originally this upstreamFiles input was filtered to be only kots manifests and I forgot to move the filter into this function during one of the rewrites 🤦